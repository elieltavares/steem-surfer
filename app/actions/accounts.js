import { encryptKey } from '../utils/crypto';
import { setItem, removeItem, getByPrefix } from '../helpers/storage';

export const ACCOUNT_ADDED = 'accounts/ACCOUNT_ADDED';
export const ACCOUNT_ADDED_SC = 'accounts/ACCOUNT_SC_ADDED';
export const ACCOUNT_UPDATED_SC = 'accounts/ACCOUNT_UPDATED_SC';
export const ACCOUNT_DELETED = 'accounts/ACCOUNT_DELETED';
export const ACCOUNTS_DELETED = 'accounts/ACCOUNTS_DELETED';

export const addAccount = (username, keys) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  // key encryption
  const eKeys = Object.assign(
    {},
    ...Object.keys(keys).map(k => ({
      [k]: keys[k] ? encryptKey(keys[k], pin) : null
    }))
  );

  const accountData = {
    type: 's',
    username,
    keys: eKeys
  };

  setItem(`account_${username}`, accountData);
  // setItem(`active_account`, username);
  dispatch(accountAdded(username, accountData));
};

export const addAccountSc = (
  username,
  accessToken,
  refreshToken,
  expiresIn,
  updated = false
) => (dispatch, getState) => {
  const { global } = getState();

  const { pin } = global;

  const accountData = {
    type: 'sc',
    username,
    accessToken: encryptKey(accessToken, pin),
    refreshToken: encryptKey(refreshToken, pin),
    expiresIn
  };

  setItem(`account_${username}`, accountData);
  if (!updated) {
    dispatch(accountAddedSc(username, accountData));
  } else {
    dispatch(accountUpdatedSc(username, accountData));
  }
};

export const deleteAccount = username => dispatch => {
  removeItem(`account_${username}`);

  dispatch(accountDeleted(username));
};

export const deleteAccounts = () => dispatch => {
  getByPrefix('account_').forEach(ac => {
    removeItem(`account_${ac.username}`);
  });

  dispatch(accountsDeleted());
};

/* action creators */

export const accountAdded = username => ({
  type: ACCOUNT_ADDED,
  payload: { username }
});

export const accountAddedSc = username => ({
  type: ACCOUNT_ADDED_SC,
  payload: { username }
});

export const accountUpdatedSc = (username, accountData) => ({
  type: ACCOUNT_UPDATED_SC,
  payload: { username, accountData }
});

export const accountDeleted = username => ({
  type: ACCOUNT_DELETED,
  payload: { username }
});

export const accountsDeleted = () => ({
  type: ACCOUNTS_DELETED,
  payload: {}
});
