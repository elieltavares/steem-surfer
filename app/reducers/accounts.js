import { getByPrefix } from '../helpers/storage';

import {
  ACCOUNT_ADDED,
  ACCOUNT_ADDED_SC,
  ACCOUNT_UPDATED_SC,
  ACCOUNT_DELETED,
  ACCOUNTS_DELETED
} from '../actions/accounts';

const defaultState = getByPrefix('account_');

export default function accounts(state = defaultState, action) {
  switch (action.type) {
    case ACCOUNT_ADDED:
    case ACCOUNT_ADDED_SC:
    case ACCOUNT_UPDATED_SC:
    case ACCOUNT_DELETED:
    case ACCOUNTS_DELETED: {
      return getByPrefix('account_');
    }
    default:
      return state;
  }
}
