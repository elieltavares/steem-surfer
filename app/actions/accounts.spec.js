/* eslint-disable */
import * as actions from './accounts';

describe('accounts actions', () => {
  it('(1) accountAdded action creator', () => {
    expect(actions.accountAdded('foo')).toMatchSnapshot();
  });

  it('(1) accountAddedSc action creator', () => {
    expect(actions.accountAddedSc('foo')).toMatchSnapshot();
  });

  it('(2) accountDeleted action creator', () => {
    expect(actions.accountDeleted('foo')).toMatchSnapshot();
  });
});
