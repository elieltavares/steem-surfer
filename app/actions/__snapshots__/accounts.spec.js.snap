// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accounts actions (1) accountAdded action creator 1`] = `
Object {
  "payload": Object {
    "username": "foo",
  },
  "type": "accounts/ACCOUNT_ADDED",
}
`;

exports[`accounts actions (1) accountAddedSc action creator 1`] = `
Object {
  "payload": Object {
    "username": "foo",
  },
  "type": "accounts/ACCOUNT_SC_ADDED",
}
`;

exports[`accounts actions (2) accountDeleted action creator 1`] = `
Object {
  "payload": Object {
    "username": "foo",
  },
  "type": "accounts/ACCOUNT_DELETED",
}
`;
